cpp/src/io/parquet/page_enc.cu

```diff
@@ -1337,11 +1345,196 @@ static __device__ void byte_reverse128(__int128_t v, void* dst)
                d_char_ptr);
 }
 
-__device__ void get_extremum(const statistics_val* stats_val,
-                             statistics_dtype dtype,
-                             void* scratch,
-                             const void** val,
-                             uint32_t* len)
+/**
+ * @brief Test to see if an array contains a valid UTF-8 string.
+ *
+ * @param str device_span containing array to test.
+ * @return true if the array contains a valid UTF-8 string.
+ */
+__device__ bool is_valid_utf8_string(device_span<uint8_t const> str)
+{
+  auto idx = 0;
+  while (idx < str.size_bytes()) {
+    // UTF-8 character should start with valid beginning byte
+    if (not strings::detail::is_valid_begin_utf8_char(str[idx])) { return false; }
+    // subsequent bytes in the characater should be continuation bytes
+    auto const width = strings::detail::bytes_in_utf8_byte(str[idx++]);
+    for (size_type i = 1; i < width && idx < str.size_bytes(); i++, idx++) {
```

Comment:
This code was in utf8.h originally, and when there I did test for that case.  But I moved it later because I didn't want to have to check an entire huge string for validity when I'm going to be truncating it anyway.  So now I only check the first `trunc_length` bytes, which may not fall on a valid UTF-8 boundary.  If the string was valid up to that point, I'm good.

Reply:
Works for me. You've thought about it and that is the most important thing.